File: BackupUtility/Classes/CommandHandler.vb
Imports GenericClassLibrary.FileSystem

Public Class CommandHandler
    Implements IFileReadableChecker, IFileIgnoreChecker, INeedToCopyChecker

    'Private ReadOnly _commands As Commands

    Private ReadOnly _fileReadableChecker As IFileReadableChecker
    Private ReadOnly ExceptionHandler As IExceptionHandler
    Private ReadOnly _fileIgnoreChecker As IFileIgnoreChecker
    Private ReadOnly _needToCopyChecker As INeedToCopyChecker

    Public Event ProcessInfoEvent(ByVal theInfo As String)

    Public Sub RaiseThisEvent(ByVal theInfo As String)
        RaiseEvent ProcessInfoEvent(theInfo)
    End Sub

    Public Sub New(commands As Commands)
        '_commands = commands

        Dim fileSystem As IFileSystem = New FileSystem()
        If commands.SkipCopyErrors Then
            ExceptionHandler = New ContinueOnExceptionHandler(Me)
        Else
            ExceptionHandler = New StopOnExceptionHandler()
        End If

        If commands.ReadCheckFirst Then
            _fileReadableChecker = New FullReadChecker(ExceptionHandler)
        Else
            _fileReadableChecker = New BasicReadChecker()
        End If
        If commands.ExclusiveExt.Length > 0 Then
            _fileIgnoreChecker = New IgnoreOnExtensionsChecker(commands.ExclusiveExt, fileSystem, ExceptionHandler)
        Else
            _fileIgnoreChecker = New IgnoreNoneChecker()
        End If
    End Sub

    Public Function CanReadFile(aFile As String) As Boolean Implements IFileReadableChecker.CanReadFile
        _fileReadableChecker.CanReadFile(aFile)
    End Function

    Public Function IgnoreFile(file As String) As Boolean Implements IFileIgnoreChecker.IgnoreFile
        _fileIgnoreChecker.IgnoreFile(file)
    End Function

    Public Function NeedToCopy(aSource As String, aTarget As String) As Boolean Implements INeedToCopyChecker.NeedToCopy
        Throw New NotImplementedException()
    End Function
End Class

File: BackupUtility/Classes/CommandHandler.cs
using GenericClassLibrary.FileSystem;
using GenericClassLibrary.Logging;
using System;
using System.IO;
using ZCopy.Classes.ExceptionHandling;
using ZCopy.Classes.FileIgnore;
using ZCopy.Classes.NeedToCopy;
using ZCopy.Classes.Readability;
using ZCopy.Interfaces;

namespace ZCopy.Classes
{
    public class CommandHandler : ICommandHandler
    {

        private readonly IFileReadableChecker _fileReadableChecker;
        private readonly IExceptionHandler _exceptionHandler;
        private readonly IFileSystem _fileSystem;
        private readonly Commands _commands;

        public event ProcessInfoEventEventHandler ProcessInfoEvent;

        public delegate void ProcessInfoEventEventHandler(object sender, ProcessInfoEventArgs eventArgs);

        public delegate bool ConfirmationRequest(string theInfo);

        public event ConfirmationRequest ConfirmationRequestHandler;

        public bool GetConfirmation(string aTarget)
        {
            if (_commands.RequestConfirm && ConfirmationRequestHandler != null)
                return ConfirmationRequestHandler.Invoke(aTarget);
            return true;
        }

        public CommandHandler(Commands commands)
        {
            _commands = commands;
            _fileSystem = new FileSystem();
            if (commands.SkipCopyErrors)
                _exceptionHandler = new ContinueOnExceptionHandler(this);
            else
                _exceptionHandler = new StopOnExceptionHandler();

            if (commands.ReadCheckFirst)
                _fileReadableChecker = new FullReadChecker(_exceptionHandler);
            else
                _fileReadableChecker = new BasicReadChecker();
        }

        public bool CanReadFile(string aFile)
        {
            return _fileReadableChecker.CanReadFile(aFile);
        }

        public bool IgnoreFile(FolderMap baseMap, string file)
        {
            IFileIgnoreChecker fileIgnoreChecker;
            if (baseMap.ExclusiveExt.Length > 0)
                fileIgnoreChecker = new IgnoreOnExtensionsChecker(baseMap.ExclusiveExt, _fileSystem, _exceptionHandler);
            else
                fileIgnoreChecker = new IgnoreNoneChecker();
            return fileIgnoreChecker.IgnoreFile(baseMap, file);
        }

        public bool NeedToCopy(FolderMap baseMap, string aSource, string aTarget)
        {
            Logger.Debug($"baseMap.UpdatedOnly = {baseMap.UpdatedOnly}");
            INeedToCopyChecker needToCopyChecker;
            if (baseMap.UpdatedOnly)
            {
                needToCopyChecker = new NeedToCopyUpdatedOnlyChecker(new FileComparer(), new FileSystem(), this);
            }
            else
            {
                needToCopyChecker = new NeedToCopyWithConfirmation(new FileSystem(), this);
            }
            return needToCopyChecker.NeedToCopy(baseMap, aSource, aTarget);
        }

        public void LogEvent(string message)
        {
            ProcessInfoEvent?.Invoke(this, new ProcessInfoEventArgs(message));
        }

        public void CopyFile(string aFile, string aTarget)
        {
            try
            {
                _fileSystem.File.Copy(aFile, aTarget, true);
            }
            catch (UnauthorizedAccessException ex)
            {
                _exceptionHandler.HandleException("Failed to copy " + aFile + " to " + aTarget + " failed(" + ex.Message + ")", ex);
            }
            catch (IOException ex)
            {
                _exceptionHandler.HandleException("Failed to copy " + aFile + " to " + aTarget + " failed(" + ex.Message + ")", ex);
            }
        }

        public bool CreateDirectory(string theTarget)
        {
            if (_fileSystem.Directory.Exists(theTarget))
            {
                return true;
            }

            bool result = false;
            // Target folder does not exist Create it
            ProcessInfoEvent?.Invoke(this, new ProcessInfoEventArgs("Create directory: " + theTarget));
            try
            {
                _fileSystem.Directory.Create(theTarget);
                result = true;
            }
            catch (DirectoryNotFoundException ex)
            {
                _exceptionHandler.HandleException("Failed to create directory " + theTarget + "(" + ex.Message + ")", ex);
                result = false;
            }
            return result;
        }

        public void HandleException(string message, Exception ex)
        {
            _exceptionHandler.HandleException(message, ex);
        }
    }
}
